model/change_point_detection.go

```diff
@@ -428,3 +428,60 @@ func createChangePoint(ctx context.Context, env cedar.Environment, resultToUpdat
 	_, err := env.GetDB().Collection(perfResultCollection).UpdateOne(ctx, filter, update)
 	return errors.Wrap(err, "Unable to create change point")
 }
+
+type ChangePointStub struct {
+	PerfResultID string `json:"perf_result_id"`
+	Measurement  string `json:"measurement"`
+}
+
+func TriageChangePoints(ctx context.Context, env cedar.Environment, changePoints []ChangePointStub, status TriageStatus) error {
+	coll := env.GetDB().Collection(perfResultCollection)
+
+	var conditions []bson.M
+	for _, stub := range changePoints {
+		conditions = append(conditions, bson.M{
+			perfIDKey: stub.PerfResultID,
+			bsonutil.GetDottedKeyName(perfAnalysisKey, perfAnalysisChangePointsKey, perfChangePointMeasurementKey): stub.Measurement,
+		})
+	}
+	filter := bson.M{
+		"$or": conditions,
+	}
+
+	cur, err := coll.Find(ctx, filter)
+	if err != nil {
+		return errors.Wrap(err, "Could not execute query finding change points for triage")
+	}
+	var results []PerformanceResult
+	if err := cur.All(ctx, &results); err != nil {
+		return errors.Wrap(err, "Could not decode performance results for triage")
+	}
+
+	for _, stub := range changePoints {
+		for _, res := range results {
+			if res.ID == stub.PerfResultID {
+				for _, cp := range res.Analysis.ChangePoints {
+					if cp.Measurement == stub.Measurement {
+						goto cont
+					}
+				}
+			}
+		}
+		return errors.Errorf("Could not find change point <%s> for performance result %s", stub.Measurement, stub.PerfResultID)
+	cont:
+		continue
+	}
+
+	update := bson.M{
+		"$set": bson.M{
+			bsonutil.GetDottedKeyName(perfAnalysisKey, perfAnalysisChangePointsKey, "$", perfChangePointTriageKey, perfTriageInfoStatusKey):    status,
+			bsonutil.GetDottedKeyName(perfAnalysisKey, perfAnalysisChangePointsKey, "$", perfChangePointTriageKey, perfTriageInfoTriagedOnKey): time.Now(),
+		},
+	}
+	for _, cond := range conditions {
+		if _, err := env.GetDB().Collection(perfResultCollection).UpdateOne(ctx, cond, update); err != nil {
```

Comment:
I can't see how to make this an updatemany.

```
replset:PRIMARY> filter
{
	"$or" : [
		{
			"_id" : "82d9f0c1a0355dc54995ff88656af1c8a3fb14f1",
			"analysis.change_points.measurement" : "measurement"
		},
		{
			"_id" : "b87a1d2a7c19c263cad984e245f7ae222e31cb29",
			"analysis.change_points.measurement" : "measurement"
		},
		{
			"_id" : "fd2cf39bbc6d68bf1c6e462586abf73d8668db29",
			"analysis.change_points.measurement" : "measurement"
		},
		{
			"_id" : "ccdb2eb4ee3da56ee2d37cc4c77136d559fd5b99",
			"analysis.change_points.measurement" : "measurement"
		},
		{
			"_id" : "a8d4723706b194924df79c47fb32a81a4a9d4b33",
			"analysis.change_points.measurement" : "measurement"
		}
	]
}
replset:PRIMARY> update
{
	"$set" : {
		"analysis.change_points.$.triage.triage_status" : "true_positive",
		"analysis.change_points.$.triage.triaged_on" : "2020-04-21T13:30:39.326068-04:00"
	}
}
replset:PRIMARY> db.perf_results.updateMany(filter, update)
2020-04-21T13:48:42.454-0400 E  QUERY    [js] WriteError({
	"index" : 0,
	"code" : 2,
	"errmsg" : "The positional operator did not find the match needed from the query.",
	"op" : {
		"q" : {
			"$or" : [
				{
					"_id" : "82d9f0c1a0355dc54995ff88656af1c8a3fb14f1",
					"analysis.change_points.measurement" : "measurement"
				},
				{
					"_id" : "b87a1d2a7c19c263cad984e245f7ae222e31cb29",
					"analysis.change_points.measurement" : "measurement"
				},
				{
					"_id" : "fd2cf39bbc6d68bf1c6e462586abf73d8668db29",
					"analysis.change_points.measurement" : "measurement"
				},
				{
					"_id" : "ccdb2eb4ee3da56ee2d37cc4c77136d559fd5b99",
					"analysis.change_points.measurement" : "measurement"
				},
				{
					"_id" : "a8d4723706b194924df79c47fb32a81a4a9d4b33",
					"analysis.change_points.measurement" : "measurement"
				}
			]
		},
		"u" : {
			"$set" : {
				"analysis.change_points.$.triage.triage_status" : "true_positive",
				"analysis.change_points.$.triage.triaged_on" : "2020-04-21T13:30:39.326068-04:00"
			}
		},
		"multi" : true,
		"upsert" : false
	}
}) :
WriteError({
	"index" : 0,
	"code" : 2,
	"errmsg" : "The positional operator did not find the match needed from the query.",
	"op" : {
		"q" : {
			"$or" : [
				{
					"_id" : "82d9f0c1a0355dc54995ff88656af1c8a3fb14f1",
					"analysis.change_points.measurement" : "measurement"
				},
				{
					"_id" : "b87a1d2a7c19c263cad984e245f7ae222e31cb29",
					"analysis.change_points.measurement" : "measurement"
				},
				{
					"_id" : "fd2cf39bbc6d68bf1c6e462586abf73d8668db29",
					"analysis.change_points.measurement" : "measurement"
				},
				{
					"_id" : "ccdb2eb4ee3da56ee2d37cc4c77136d559fd5b99",
					"analysis.change_points.measurement" : "measurement"
				},
				{
					"_id" : "a8d4723706b194924df79c47fb32a81a4a9d4b33",
					"analysis.change_points.measurement" : "measurement"
				}
			]
		},
		"u" : {
			"$set" : {
				"analysis.change_points.$.triage.triage_status" : "true_positive",
				"analysis.change_points.$.triage.triaged_on" : "2020-04-21T13:30:39.326068-04:00"
			}
		},
		"multi" : true,
		"upsert" : false
	}
})
WriteError@src/mongo/shell/bulk_api.js:458:48
mergeBatchResults@src/mongo/shell/bulk_api.js:855:49
executeBatch@src/mongo/shell/bulk_api.js:919:13
Bulk/this.execute@src/mongo/shell/bulk_api.js:1163:21
DBCollection.prototype.updateMany@src/mongo/shell/crud_api.js:666:17
@(shell):1:1
replset:PRIMARY> db.perf_results.update(filter["$or"][0], update)
WriteResult({ "nMatched" : 1, "nUpserted" : 0, "nModified" : 1 })
```

Reply:
If we can't do an updatemany, I think we might need to revert back to the transactional implementation.

Reply:
you can do a BulkWrite api with many UpdateOnes  (ordered=true) which isn't trasnactional, but will do the thing you want in a single network round trip

I think it's safe to avoid the transaction if the system can recover for a partially applied updatemany